File: bindings/php/src/ProvenSshBastion.php
<?php

// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// PHP bindings for the proven-ssh-bastion Zig FFI.

declare(strict_types=1);

namespace ProvenServers;

/** SSH bastion session states matching Idris2 ABI tags. */
enum SshBastionState: int
{
    case Connected    = 0;
    case KeyExchanged = 1;
    case Authenticated = 2;
    case ChannelOpen  = 3;
    case Active       = 4;
    case Closed       = 5;
}

/** SSH key exchange methods matching Idris2 ABI tags. */
enum SshKexMethod: int
{
    case Curve25519Sha256            = 0;
    case EcdhSha2Nistp256            = 1;
    case EcdhSha2Nistp384            = 2;
    case DiffieHellmanGroup14Sha256  = 3;
    case DiffieHellmanGroup16Sha512  = 4;
}

/** SSH authentication methods matching Idris2 ABI tags. */
enum SshAuthMethod: int
{
    case PublicKey            = 0;
    case Password             = 1;
    case KeyboardInteractive  = 2;
    case HostBased            = 3;
}

/** SSH channel types matching Idris2 ABI tags. */
enum SshChannelType: int
{
    case Session        = 0;
    case DirectTcpip    = 1;
    case ForwardedTcpip = 2;
    case X11            = 3;
}

/** SSH channel states matching Idris2 ABI tags. */
enum SshChannelState: int
{
    case Opening = 0;
    case Open    = 1;
    case Closed  = 2;
}

/** SSH disconnect reasons matching Idris2 ABI tags. */
enum SshDisconnectReason: int
{
    case ByApplication        = 0;
    case ProtocolError        = 1;
    case KeyExchangeFailed    = 2;
    case AuthCancelledByUser  = 3;
    case TooManyConnections   = 4;
    case HostNotAllowed       = 5;
    case IllegalUserName      = 6;
}

/**
 * SSH bastion session context wrapping a Zig FFI slot.
 */
final class ProvenSshBastion
{
    private const CDEF = <<<'CDEF'
    int ssh_bastion_create(uint8_t kex, uint8_t auth);
    void ssh_bastion_destroy(int slot);
    uint8_t ssh_bastion_state(int slot);
    uint8_t ssh_bastion_kex_method(int slot);
    uint8_t ssh_bastion_auth_method(int slot);
    uint8_t ssh_bastion_can_transfer(int slot);
    uint8_t ssh_bastion_disconnect_reason(int slot);
    uint32_t ssh_bastion_auth_failures(int slot);
    uint8_t ssh_bastion_complete_kex(int slot);
    uint8_t ssh_bastion_authenticate(int slot, uint8_t dummy);
    uint8_t ssh_bastion_record_auth_failure(int slot);
    int ssh_bastion_open_channel(int slot, uint8_t ch_type);
    uint8_t ssh_bastion_confirm_channel(int slot, int ch_id);
    uint8_t ssh_bastion_close_channel(int slot, int ch_id);
    uint8_t ssh_bastion_channel_state(int slot, int ch_id);
    uint8_t ssh_bastion_channel_type(int slot, int ch_id);
    uint32_t ssh_bastion_channel_count(int slot);
    uint8_t ssh_bastion_rekey(int slot);
    uint8_t ssh_bastion_disconnect(int slot, uint8_t reason);
    uint32_t ssh_bastion_audit_count(int slot);
    uint8_t ssh_bastion_audit_entry(int slot, uint32_t index);
    uint8_t ssh_bastion_audit_entry_to(int slot, uint32_t index);
    uint8_t ssh_bastion_set_recording(int slot, uint8_t enabled);
    uint8_t ssh_bastion_is_recording(int slot);
    uint32_t ssh_bastion_abi_version(void);
    uint8_t ssh_bastion_can_transition(uint8_t from, uint8_t to);
    CDEF;

    private static ?\FFI $ffi = null;
    private int $slot;
    private bool $destroyed = false;

    private function __construct(int $slot) { $this->slot = $slot; }

    private static function ffi(): \FFI
    {
        if (self::$ffi === null) {
            self::$ffi = ProvenServers::loadLibrary('ssh_bastion', self::CDEF);
        }
        return self::$ffi;
    }

    /** @throws ProvenError */
    public static function create(
        SshKexMethod $kex = SshKexMethod::Curve25519Sha256,
        SshAuthMethod $auth = SshAuthMethod::PublicKey,
    ): self {
        return new self(ProvenError::checkSlot(self::ffi()->ssh_bastion_create($kex->value, $auth->value)));
    }

    public function destroy(): void
    {
        if (!$this->destroyed) {
            self::ffi()->ssh_bastion_destroy($this->slot);
            $this->destroyed = true;
        }
    }

    public function state(): ?SshBastionState
    {
        $tag = self::ffi()->ssh_bastion_state($this->slot);
        return $tag <= 5 ? SshBastionState::from($tag) : null;
    }

    public function kexMethod(): ?SshKexMethod
    {
        $tag = self::ffi()->ssh_bastion_kex_method($this->slot);
        return $tag <= 4 ? SshKexMethod::from($tag) : null;
    }

    public function authMethod(): ?SshAuthMethod
    {
        $tag = self::ffi()->ssh_bastion_auth_method($this->slot);
        return $tag <= 3 ? SshAuthMethod::from($tag) : null;
    }

    public function canTransferData(): bool { return self::ffi()->ssh_bastion_can_transfer($this->slot) === 1; }

    public function disconnectReason(): ?SshDisconnectReason
    {
        $tag = self::ffi()->ssh_bastion_disconnect_reason($this->slot);
        return $tag <= 6 ? SshDisconnectReason::from($tag) : null;
    }

    public function authFailures(): int { return self::ffi()->ssh_bastion_auth_failures($this->slot); }

    /** @throws ProvenError */
    public function completeKex(): void { ProvenError::checkStatus(self::ffi()->ssh_bastion_complete_kex($this->slot)); }
    /** @throws ProvenError */
    public function authenticate(): void { ProvenError::checkStatus(self::ffi()->ssh_bastion_authenticate($this->slot, 0)); }

    /** @return bool True if locked out. */
    public function recordAuthFailure(): bool { return self::ffi()->ssh_bastion_record_auth_failure($this->slot) === 1; }

    /** @throws ProvenError @return int Channel ID. */
    public function openChannel(SshChannelType $chType): int
    {
        return ProvenError::checkSlot(self::ffi()->ssh_bastion_open_channel($this->slot, $chType->value));
    }

    /** @throws ProvenError */
    public function confirmChannel(int $chId): void
    {
        ProvenError::checkStatus(self::ffi()->ssh_bastion_confirm_channel($this->slot, $chId));
    }

    /** @throws ProvenError */
    public function closeChannel(int $chId): void
    {
        ProvenError::checkStatus(self::ffi()->ssh_bastion_close_channel($this->slot, $chId));
    }

    public function channelState(int $chId): ?SshChannelState
    {
        $tag = self::ffi()->ssh_bastion_channel_state($this->slot, $chId);
        return $tag <= 2 ? SshChannelState::from($tag) : null;
    }

    public function channelType(int $chId): ?SshChannelType
    {
        $tag = self::ffi()->ssh_bastion_channel_type($this->slot, $chId);
        return $tag <= 3 ? SshChannelType::from($tag) : null;
    }

    public function channelCount(): int { return self::ffi()->ssh_bastion_channel_count($this->slot); }

    /** @throws ProvenError */
    public function rekey(): void { ProvenError::checkStatus(self::ffi()->ssh_bastion_rekey($this->slot)); }

    /** @throws ProvenError */
    public function disconnect(SshDisconnectReason $reason): void
    {
        ProvenError::checkStatus(self::ffi()->ssh_bastion_disconnect($this->slot, $reason->value));
    }

    public function auditCount(): int { return self::ffi()->ssh_bastion_audit_count($this->slot); }

    public function auditEntryFrom(int $index): ?SshBastionState
    {
        $tag = self::ffi()->ssh_bastion_audit_entry($this->slot, $index);
        return $tag <= 5 ? SshBastionState::from($tag) : null;
    }

    public function auditEntryTo(int $index): ?SshBastionState
    {
        $tag = self::ffi()->ssh_bastion_audit_entry_to($this->slot, $index);
        return $tag <= 5 ? SshBastionState::from($tag) : null;
    }

    /** @throws ProvenError */
    public function setRecording(bool $enabled): void
    {
        ProvenError::checkStatus(self::ffi()->ssh_bastion_set_recording($this->slot, $enabled ? 1 : 0));
    }

    public function isRecording(): bool { return self::ffi()->ssh_bastion_is_recording($this->slot) === 1; }

    public static function abiVersion(): int { return self::ffi()->ssh_bastion_abi_version(); }
    public static function canTransition(SshBastionState $from, SshBastionState $to): bool
    {
        return self::ffi()->ssh_bastion_can_transition($from->value, $to->value) === 1;
    }
}

File: bindings/ocaml/lib/proven_ssh_bastion.ml
(* SPDX-License-Identifier: MPL-2.0 *)
(* Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk> *)

(** SSH Bastion protocol bindings for proven-servers.

    Wraps the C-ABI functions from
    [protocols/proven-ssh-bastion/ffi/zig/src/ssh_bastion.zig]. *)

(** SSH bastion lifecycle states. *)
type bastion_state =
  | Connected      (** TCP connection established. *)
  | Key_exchanged  (** Key exchange completed. *)
  | Authenticated  (** User authenticated. *)
  | Channel_open   (** At least one channel opening. *)
  | Active         (** Session fully active. *)
  | Closed         (** Session closed. *)

(** Key exchange methods. *)
type kex_method =
  | Curve25519 | Dh_group14 | Dh_group16 | Ecdh_p256 | Ecdh_p384

(** Authentication methods. *)
type auth_method =
  | Public_key | Password | Keyboard | Certificate

(** SSH channel types. *)
type channel_type =
  | Session | Direct_tcp_ip | Forwarded_tcp_ip | Subsystem

(** SSH channel states. *)
type channel_state =
  | Ch_opening | Ch_open | Ch_closing | Ch_closed

(** SSH disconnect reasons (RFC 4253). *)
type disconnect_reason =
  | Host_not_allowed | Protocol_error | Key_exchange_failed
  | Auth_failed | Service_not_available | By_application
  | Too_many_connections

let bastion_state_to_tag = function
  | Connected -> 0 | Key_exchanged -> 1 | Authenticated -> 2
  | Channel_open -> 3 | Active -> 4 | Closed -> 5

let bastion_state_of_tag = function
  | 0 -> Some Connected | 1 -> Some Key_exchanged | 2 -> Some Authenticated
  | 3 -> Some Channel_open | 4 -> Some Active | 5 -> Some Closed
  | _ -> None

let kex_method_to_tag = function
  | Curve25519 -> 0 | Dh_group14 -> 1 | Dh_group16 -> 2
  | Ecdh_p256 -> 3 | Ecdh_p384 -> 4

let auth_method_to_tag = function
  | Public_key -> 0 | Password -> 1 | Keyboard -> 2 | Certificate -> 3

let auth_method_of_tag = function
  | 0 -> Some Public_key | 1 -> Some Password | 2 -> Some Keyboard
  | 3 -> Some Certificate | _ -> None

let channel_type_to_tag = function
  | Session -> 0 | Direct_tcp_ip -> 1 | Forwarded_tcp_ip -> 2 | Subsystem -> 3

let channel_state_of_tag = function
  | 0 -> Some Ch_opening | 1 -> Some Ch_open | 2 -> Some Ch_closing
  | 3 -> Some Ch_closed | _ -> None

let channel_type_of_tag = function
  | 0 -> Some Session | 1 -> Some Direct_tcp_ip | 2 -> Some Forwarded_tcp_ip
  | 3 -> Some Subsystem | _ -> None

let disconnect_reason_to_tag = function
  | Host_not_allowed -> 0 | Protocol_error -> 1 | Key_exchange_failed -> 2
  | Auth_failed -> 3 | Service_not_available -> 4 | By_application -> 5
  | Too_many_connections -> 6

let disconnect_reason_of_tag = function
  | 0 -> Some Host_not_allowed | 1 -> Some Protocol_error
  | 2 -> Some Key_exchange_failed | 3 -> Some Auth_failed
  | 4 -> Some Service_not_available | 5 -> Some By_application
  | 6 -> Some Too_many_connections | _ -> None

(* --- C FFI declarations --- *)

external c_ssh_bastion_abi_version : unit -> int = "ssh_bastion_abi_version"
external c_ssh_bastion_create : int -> int -> int = "ssh_bastion_create"
external c_ssh_bastion_destroy : int -> unit = "ssh_bastion_destroy"
external c_ssh_bastion_state : int -> int = "ssh_bastion_state"
external c_ssh_bastion_kex_method : int -> int = "ssh_bastion_kex_method"
external c_ssh_bastion_auth_method : int -> int = "ssh_bastion_auth_method"
external c_ssh_bastion_can_transfer : int -> int = "ssh_bastion_can_transfer"
external c_ssh_bastion_disconnect_reason : int -> int = "ssh_bastion_disconnect_reason"
external c_ssh_bastion_auth_failures : int -> int = "ssh_bastion_auth_failures"
external c_ssh_bastion_complete_kex : int -> int = "ssh_bastion_complete_kex"
external c_ssh_bastion_authenticate : int -> int -> int = "ssh_bastion_authenticate"
external c_ssh_bastion_record_auth_failure : int -> int = "ssh_bastion_record_auth_failure"
external c_ssh_bastion_open_channel : int -> int -> int = "ssh_bastion_open_channel"
external c_ssh_bastion_confirm_channel : int -> int -> int = "ssh_bastion_confirm_channel"
external c_ssh_bastion_close_channel : int -> int -> int = "ssh_bastion_close_channel"
external c_ssh_bastion_channel_state : int -> int -> int = "ssh_bastion_channel_state"
external c_ssh_bastion_channel_type : int -> int -> int = "ssh_bastion_channel_type"
external c_ssh_bastion_channel_count : int -> int = "ssh_bastion_channel_count"
external c_ssh_bastion_rekey : int -> int = "ssh_bastion_rekey"
external c_ssh_bastion_disconnect : int -> int -> int = "ssh_bastion_disconnect"
external c_ssh_bastion_can_transition : int -> int -> int = "ssh_bastion_can_transition"
external c_ssh_bastion_audit_count : int -> int = "ssh_bastion_audit_count"
external c_ssh_bastion_set_recording : int -> int -> int = "ssh_bastion_set_recording"
external c_ssh_bastion_is_recording : int -> int = "ssh_bastion_is_recording"

(* --- Safe wrappers --- *)

let abi_version () = c_ssh_bastion_abi_version ()

let create ~kex ~auth =
  Proven_error.from_slot
    (c_ssh_bastion_create (kex_method_to_tag kex) (auth_method_to_tag auth))

let destroy slot = c_ssh_bastion_destroy slot
let get_state slot = bastion_state_of_tag (c_ssh_bastion_state slot)

let get_kex_method slot =
  let tag = c_ssh_bastion_kex_method slot in
  if tag <= 4 then Some (match tag with
    | 0 -> Curve25519 | 1 -> Dh_group14 | 2 -> Dh_group16
    | 3 -> Ecdh_p256 | _ -> Ecdh_p384)
  else None

let get_auth_method slot = auth_method_of_tag (c_ssh_bastion_auth_method slot)
let can_transfer_data slot = c_ssh_bastion_can_transfer slot = 1
let get_disconnect_reason slot = disconnect_reason_of_tag (c_ssh_bastion_disconnect_reason slot)
let auth_failures slot = c_ssh_bastion_auth_failures slot
let complete_kex slot = Proven_error.from_status (c_ssh_bastion_complete_kex slot)
let authenticate slot = Proven_error.from_status (c_ssh_bastion_authenticate slot 0)
let record_auth_failure slot = c_ssh_bastion_record_auth_failure slot = 1

let open_channel slot ch_type =
  Proven_error.from_slot (c_ssh_bastion_open_channel slot (channel_type_to_tag ch_type))

let confirm_channel slot ch_id =
  Proven_error.from_status (c_ssh_bastion_confirm_channel slot ch_id)

let close_channel slot ch_id =
  Proven_error.from_status (c_ssh_bastion_close_channel slot ch_id)

let channel_state slot ch_id =
  channel_state_of_tag (c_ssh_bastion_channel_state slot ch_id)

let channel_type slot ch_id =
  channel_type_of_tag (c_ssh_bastion_channel_type slot ch_id)

let channel_count slot = c_ssh_bastion_channel_count slot
let rekey slot = Proven_error.from_status (c_ssh_bastion_rekey slot)

let disconnect slot reason =
  Proven_error.from_status (c_ssh_bastion_disconnect slot (disconnect_reason_to_tag reason))

let can_transition ~from ~to_ =
  c_ssh_bastion_can_transition (bastion_state_to_tag from) (bastion_state_to_tag to_) = 1

let audit_count slot = c_ssh_bastion_audit_count slot

let set_recording slot ~enabled =
  Proven_error.from_status (c_ssh_bastion_set_recording slot (if enabled then 1 else 0))

let is_recording slot = c_ssh_bastion_is_recording slot = 1
